pkg/app/api/grpcapi/web_api.go

```diff
@@ -194,6 +197,73 @@ func (a *WebAPI) DeleteEnvironment(ctx context.Context, req *webservice.DeleteEn
 	if err := a.validateEnvBelongsToProject(ctx, req.EnvironmentId, claims.Role.ProjectId); err != nil {
 		return nil, err
 	}
+	// Check if no Piped has permission to the given environment.
+	pipeds, err := a.pipedStore.ListPipeds(ctx, datastore.ListOptions{
+		Filters: []datastore.ListFilter{
+			{
+				Field:    "EnvIds",
+				Operator: datastore.OperatorIn,
+				Value:    req.EnvironmentId,
+			},
+		},
+	})
+	if err != nil {
+		a.logger.Error("failed to fetch Pipeds linked to the given environment",
+			zap.String("env-id", req.EnvironmentId),
+			zap.Error(err),
+		)
+		return nil, status.Error(codes.Internal, "Failed to validate the deletion operation")
+	}
+	if len(pipeds) > 0 {
+		var pipedNames strings.Builder
+		for _, p := range pipeds {
+			pipedNames.WriteString(fmt.Sprintf("%q", p.Name))
+		}
+		return nil, status.Errorf(
+			codes.FailedPrecondition,
+			"Found Pipeds linked the environment to be deleted. Please remove this environment from all Pipeds (%s) on the Piped settings page",
+			pipedNames,
+		)
+	}
+
+	// Delete all applications that belongs to the given env.
+	apps, _, err := a.applicationStore.ListApplications(ctx, datastore.ListOptions{
+		Filters: []datastore.ListFilter{
+			{
+				Field:    "EnvId",
+				Operator: datastore.OperatorEqual,
+				Value:    req.EnvironmentId,
+			},
+		},
+	})
+	if err != nil {
+		a.logger.Error("failed to fetch applications that belongs to the given environment",
+			zap.String("env-id", req.EnvironmentId),
+			zap.Error(err),
+		)
+		return nil, status.Error(codes.Internal, "Failed to fetch applications that belongs to the given environment")
+	}
+	for _, app := range apps {
+		if err := a.validateAppBelongsToProject(ctx, app.Id, claims.Role.ProjectId); err != nil {
```

Comment:
The application model is already containing the project ID so we can validate it directly by comparing `app.ProjectId` and `claims.Role.ProjectId`) instead of using cache or making a new request.

Reply:
For sure